### ISSUE TITLE
Getting an instance of a service (DI) in Job.Execute

### ISSUE THREAD

**Issue by whvanderpost** [quartznet/quartznet#700]
Hi, I have a question.
I'm working on a ASP.Net Core 2.1(.5) project and I'm using Quartz to trigger code every x time. I have registered a service (as a singleton) in the Startup -> ConfigureServices method and I'm wondering how I can get a handle to that service from within my jobs' Execute method.

Does someone have a solution for this?

**Comment by Gprsk** [comment]
You will need to implement your own JobFactory and use it.

By implementing the "NewJob" method, you have full control of how the job is instanciated. This way you can attach/ask your DI container to the IJob or do anything you may want.

A simple example, assuming you built the JobFactory injecting an IServiceProvider:

```csharp
public IJob NewJob(TriggerFiredBundle bundle, IScheduler scheduler)
{
    try
    {
        IJobDetail jDetail = bundle.JobDetail;
        Type jobType = jDetail.JobType;

        IJob jobInstance = [My_IServiceProvider_Instance].GetService(jobType) as IJob;

        return jobInstance ?? throw new Exception("Invalid Job"); //This might stop Quartz from triggering again and should not happen
    }
    catch (Exception ex)
    {
        throw new SchedulerException($"Failed to create job '{bundle.JobDetail.JobType.FullName}'", ex);
    }
}
```

**Comment by whvanderpost** [comment]
Thanks, this is really helpful.
I've created the following class that creates the job instance and supplies the ServiceScopeFactory to the job:

```
public class TestsJobFactory : IJobFactory
{
    private readonly IServiceScopeFactory _scopeFactory;

    public TestsJobFactory(IServiceScopeFactory scopeFactory)
    {
      _scopeFactory = scopeFactory;
    }

    public IJob NewJob(TriggerFiredBundle bundle, IScheduler scheduler)
    {
      try
      {
        IJobDetail jDetail = bundle.JobDetail;
        Type jobType = jDetail.JobType;

        var jobInstance = ObjectUtils.InstantiateType<IJob>(jobType);
        ((TestJob)jobInstance).ScopeFactory = _scopeFactory;

        return jobInstance;
      }
      catch (Exception ex)
      {
        throw new SchedulerException($"Failed to create job '{bundle.JobDetail.JobType.FullName}'", ex);
      }
    }

    public void ReturnJob(IJob job)
    {
      // do nothing
    }
}
```

The job can then use the scope factory to get, for example, the db context.
```
using (var scope = ScopeFactory.CreateScope())
{
  var dbContext = scope.ServiceProvider.GetRequiredService<ApplicationDbContext>();
}
```

**Comment by lahma** [comment]
https://www.quartz-scheduler.net/2020/07/08/quartznet-3-1-beta-1-released/ describes how 3.1 beta 1 now has built-in support and examples.
